src/fireedge/src/modules/utils/parser/vmTemplateFilter.js

```diff
@@ -0,0 +1,810 @@
+    : alwaysIncludeAttributes
+
+  // Filter data of formData.general
+  const newGeneral = reduceGeneral(
+    formData?.general,
+    modifiedFields?.general,
+    normalizedTemplate?.general,
+    includeAtributes
+  )
+
+  // Filter data of formData.extra
+  const newExtra = reduceExtra(
+    formData,
+    modifiedFields,
+    normalizedTemplate,
+    tabFormMap,
+    includeAtributes
+  )
+
+  // Add custom variables
+  const newCustomVariables = {
+    ...normalizedTemplate['custom-variables'],
+    ...formData['custom-variables'],
+  }
+
+  const result = {
+    ...newGeneral,
+    ...newExtra,
+    ...newCustomVariables,
+  }
+
+  // Instantiate form could have another step called user_inputs
+  if (formData.user_inputs) {
+    result.CONTEXT = {
+      ...result.CONTEXT,
+      ...formData.user_inputs,
+    }
+  }
+
+  // Return object with all sections
+  return result
+}
+
+/**
+ * Generates a formData format object with the original data before the user modifies anything.
+ *
+ * @param {object} existingTemplate - Original data before user's modifications in the same format as formData
+ * @param {string} tabFormMap - Map with the attributes (for the XML request) that will contain the section
+ * @returns {object} - A form data object with the original data
+ */
+const normalizeTemplate = (existingTemplate, tabFormMap) => {
+  const categoryLookup = Object.entries(tabFormMap).reduce(
+    (acc, [category, keys]) => {
+      keys.forEach((key) => (acc[key] = category))
+
+      return acc
+    },
+    {}
+  )
+
+  const normalized = { general: {}, extra: {} }
+
+  Object.entries(existingTemplate).forEach(([key, value]) => {
+    if (categoryLookup[key]) {
+      normalized.extra[key] = value
+    } else {
+      normalized.general[key] = value
+    }
+  })
+
+  if ('custom-variables' in existingTemplate) {
+    normalized['custom-variables'] = existingTemplate['custom-variables']
+  }
+
+  return normalized
+}
+
+/**
+ * Filter the formData.general section to only add modifications of the user.
+ *
+ * @param {object} formData - The data from the form (the one that the user could be modify)
+ * @param {object} itemModifications - Map with the fields that will be touched and modified by the user or deleted (in array case) by the user
+ * @param {object} existingTemplate - Original data before user's modifications in the same format as formData
+ * @param {object} alwaysInclude - Include this field always as modified
+ * @returns {object} - The general section with the final data to submit
+ */
+const reduceGeneral = (
+  formData,
+  itemModifications,
+  existingTemplate,
+  alwaysInclude
+) => {
+  const newGeneral = { ...existingTemplate }
+
+  // Add always include
+  const correctionMap = merge({}, itemModifications, alwaysInclude?.general)
+
+  Object.entries(correctionMap || {}).forEach(([key, correction]) => {
+    if (correction && typeof correction === 'object' && correction.__delete__) {
+      delete newGeneral[key]
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

We detected that you are using `delete` to remove items from an array. If we use `delete`, it will delete the object property, but will not reindex the array or update its length. It leaves `undefined` holes in the array. Therefore, we recommend you to use `splice` to remove items from an array.

[Learn more](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete)